File: packages/prov_wallet_flutter/android/src/main/kotlin/com/figure/prov_wallet_flutter/ProvWalletFlutterPlugin.kt
package com.figure.prov_wallet_flutter

import android.content.Context
import android.hardware.biometrics.BiometricManager
import android.security.keystore.UserNotAuthenticatedException
import android.util.Log
import androidx.annotation.NonNull
import androidx.preference.PreferenceManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.*

class Error {
    companion object {
        const val userNotAuthenticated = "user_not_authenticated"
        const val unknown = "unknown"
    }
}

class ProvWalletFlutterPlugin: FlutterPlugin, MethodCallHandler {
    private companion object {
        val TAG = "$ProvWalletFlutterPlugin"
    }

    private lateinit var channel : MethodChannel
    private lateinit var context: Context
    private lateinit var cipherService: CipherService
    private lateinit var systemService: SystemService
    private lateinit var mainScope: CoroutineScope

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val name = "prov_wallet_flutter"
        mainScope = MainScope()
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, name)
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext

        val packageManager = context.packageManager
        val biometricManager = context.getSystemService(BiometricManager::class.java)
        val executor = context.mainExecutor
        systemService = SystemService(packageManager, biometricManager, executor)

        val preferences = PreferenceManager.getDefaultSharedPreferences(context)
        cipherService = CipherService(preferences, systemService)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        mainScope.launch {
            runCatching {
                when (call.method) {
                    "getPlatformVersion" -> {
                        result.success("Android ${android.os.Build.VERSION.RELEASE}")
                    }
                    "getBiometryType" -> {
                        val type = systemService.getBiometryType()
                        result.success(type.value)
                    }
                    "authenticateBiometry" -> {
                        val success = systemService.authenticate(context, useBiometry = true)
                        result.success(success)
                    }
                    "resetAuth" -> {
                        // Only timeout is supported on Android
                        result.success(true)
                    }
                    "encryptKey" -> {
                        val id = call.argument<String>("id")!!
                        val privateKey = call.argument<String>("private_key")!!
                        val success = cipherService.encryptKey(context, id, privateKey)
                        result.success(success)
                    }
                    "decryptKey" -> {
                        val id = call.argument<String>("id")!!

                        val decrypted = cipherService.decryptKey(context, id)

                        result.success(decrypted)
                    }
                    "removeKey" -> {
                        val id = call.argument<String>("id")!!
                        val success = cipherService.removeKey(id)
                        result.success(success)
                    }
                    "resetKeys" -> {
                        val success = cipherService.clearPin()
                        result.success(success)
                    }
                    "getUseBiometry" -> {
                        val useBiometry = cipherService.getUseBiometry()
                        result.success(useBiometry)
                    }
                    "setUseBiometry" -> {
                        val useBiometry = call.argument<Boolean>("use_biometry")!!
                        val success = cipherService.setUseBiometry(useBiometry)
                        result.success(success)
                    }
                    "getPin" -> {
                        val pin = cipherService.getPin()
                        result.success(pin)
                    }
                    "setPin" -> {
                        val pin = call.argument<String>("pin")!!
                        val success = cipherService.setPin(pin)
                        result.success(success)
                    }
                    "deletePin" -> {
                        val success = cipherService.deletePin()
                        result.success(success)
                    }
                    else -> {
                        result.notImplemented()
                    }
                }
            }.onFailure { e ->
                val stringWriter = StringWriter()
                val printWriter = PrintWriter(stringWriter)
                e.printStackTrace(printWriter)

                var details = e.toString()

                Log.e(TAG, "Exception occurred", e)

                when (e) {
                    is UserNotAuthenticatedException -> {
                        result.error(Error.userNotAuthenticated, e.message, details)
                    }
                    else -> {
                        result.error(Error.unknown, e.message, details)
                    }
                }
            }
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        mainScope.cancel()
    }
}

File: packages/prov_wallet_flutter/macos/Classes/ProvWalletFlutterPlugin.swift
import Cocoa
import FlutterMacOS

public class ProvWalletFlutterPlugin: NSObject, FlutterPlugin {
    var flutterChannel: FlutterMethodChannel?
    let cipherService: CipherService
    let upgradeTask: Task<Void, Error>

    override public init() {
        let service = CipherService()
        cipherService = service
        upgradeTask = Task {
            try await service.upgrade()
        }
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "prov_wallet_flutter", binaryMessenger: registrar.messenger)
        let instance = ProvWalletFlutterPlugin()
        instance.flutterChannel = channel

        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            var value: Any?

            do {
                try await upgradeTask.value

                if call.method == "getPlatformVersion" {
                    value = "macOS " + ProcessInfo.processInfo.operatingSystemVersionString
                } else if call.method == "getBiometryType" {
                    let type = cipherService.biometryType
                    value = type.rawValue
                } else if call.method == "authenticateBiometry" {
                    value = await cipherService.authenticate()
                } else if call.method == "getLockScreenEnabled" {
                    value = cipherService.getLockScreenEnabled()
                } else if call.method == "resetAuth" {
                    value = cipherService.resetAuth()
                } else if call.method == "encryptKey" {
                    let argsFormatted = call.arguments as? [String: Any]
                    guard let id = argsFormatted?["id"] as? String else {
                        throw CipherServiceError(kind: .invalidArgument, message: "id is required")
                    }
                    guard let privateKey = argsFormatted?["private_key"] as? String else {
                        throw CipherServiceError(kind: .invalidArgument, message: "privateKey is required")
                    }

                    try await cipherService.encryptKey(id: id, plainText: privateKey)

                    value = true
                } else if call.method == "decryptKey" {
                    let argsFormatted = call.arguments as? [String: Any]
                    guard let id = argsFormatted?["id"] as? String else {
                        throw CipherServiceError(kind: .invalidArgument, message: "id is required")
                    }

                    value = try await cipherService.decryptKey(id: id)
                } else if call.method == "removeKey" {
                    let argsFormatted = call.arguments as? [String: Any]
                    guard let id = argsFormatted?["id"] as? String else {
                        throw CipherServiceError(kind: .invalidArgument, message: "id is required")
                    }

                    value = try cipherService.removeKey(id: id)
                } else if call.method == "resetKeys" {
                    value = cipherService.resetKeys()
                } else if call.method == "getUseBiometry" {
                    value = cipherService.getUseBiometry()
                } else if call.method == "setUseBiometry" {
                    let argsFormatted = call.arguments as? [String: Any]
                    let useBiometry = argsFormatted?["use_biometry"] as? Bool ?? true

                    cipherService.setUseBiometry(useBiometry: useBiometry)
                    value = true
                } else if call.method == "getPin" {
                    value = cipherService.getPin()
                } else if call.method == "setPin" {
                    let argsFormatted = call.arguments as? [String: Any]
                    guard let pin = argsFormatted?["pin"] as? String else {
                        throw CipherServiceError(kind: .invalidArgument, message: "pin is required")
                    }

                    value = try cipherService.setPin(pin: pin)
                } else if call.method == "deletePin" {
                    value = cipherService.deletePin()
                }
            } catch {
                Utilities.log("Plugin error: \(error)")
                value = getFlutterError(error: error)
            }

            let ret = value

            await MainActor.run {
                result(ret)
            }
        }
    }

    private func getFlutterError(error: Error) -> FlutterError {
        var code: String
        var message: String
        var details: Error?

        if error is CipherServiceError {
            let provError = error as! CipherServiceError
            code = String(describing: provError.kind)
            message = provError.message
        } else {
            code = String(describing: CipherServiceError.CipherServiceErrorCode.unknown)
            message = error.localizedDescription
            details = error
        }

        return FlutterError(code: code, message: message, details: details)
    }
}
